java/ql/src/semmle/code/java/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -193,6 +193,18 @@ predicate readStep(Node node1, Content f, Node node2) {
     fr.getField() = f.(FieldContent).getField() and
     fr = node2.asExpr()
   )
+  or
+  exists(Record r, Method getter, Field recf, MethodAccess get |
+    getter.getDeclaringType() = r and
+    recf.getDeclaringType() = r and
+    getter.getNumberOfParameters() = 0 and
+    getter.getName() = recf.getName() and
+    not exists(getter.getBody()) and
```

Comment:
I suppose `abstract` records are not allowed, so checking for the absence of a body is fine, however, wouldn't it be better to use  `getter.fromSource()` instead

Reply:
Records are implicitly final and therefore can't be abstract. It looks like checking `fromSource()` would work as well, but I didn't expect it based on the qldoc of `fromSource()` - it states:
```
   * Holds if this element pertains to a source file.
   *
   * Elements pertaining to source files may include generated elements
```
So I would actually expect `fromSource()` to hold for a generated getter when the corresponding record is `fromSource()`. I wonder why that isn't the case.

Reply:
Ok, sounds like it's best to keep using the empty body check.